ports/samd/samd_qspiflash.c

```diff
@@ -0,0 +1,501 @@
+
+        addr += size;
+        pos += size;
+    }
+
+    return mp_const_none;
+}
+
+STATIC mp_obj_t samd_qspiflash_erase(uint32_t addr) {
+    wait_for_flash_ready();
+    write_enable();
+    erase_command(QSPI_CMD_ERASE_SECTOR, addr);
+
+    return mp_const_none;
+}
+
+STATIC mp_obj_t samd_qspiflash_readblocks(size_t n_args, const mp_obj_t *args) {
+    samd_qspiflash_obj_t *self = MP_OBJ_TO_PTR(args[0]);
+    uint32_t offset = (mp_obj_get_int(args[1]) * self->sectorsize);
+    mp_buffer_info_t bufinfo;
+    mp_get_buffer_raise(args[2], &bufinfo, MP_BUFFER_WRITE);
+    if (n_args == 4) {
+        offset += mp_obj_get_int(args[3]);
+    }
+
+    // Read data to flash (adf4 API)
+    samd_qspiflash_read(self, offset, bufinfo.buf, bufinfo.len);
+
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(samd_qspiflash_readblocks_obj, 3, 4, samd_qspiflash_readblocks);
+
+STATIC mp_obj_t samd_qspiflash_writeblocks(size_t n_args, const mp_obj_t *args) {
+    samd_qspiflash_obj_t *self = MP_OBJ_TO_PTR(args[0]);
+    uint32_t offset = (mp_obj_get_int(args[1]) * self->sectorsize);
+    mp_buffer_info_t bufinfo;
+    mp_get_buffer_raise(args[2], &bufinfo, MP_BUFFER_READ);
+    if (n_args == 3) {
+        samd_qspiflash_erase(offset);
+        // TODO check return value
+    } else {
+        offset += mp_obj_get_int(args[3]);
+    }
+    // Write data to flash (adf4 API)
+    samd_qspiflash_write(self,  offset, bufinfo.buf, bufinfo.len);
+    // TODO check return value
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(samd_qspiflash_writeblocks_obj, 3, 4, samd_qspiflash_writeblocks);
+
+STATIC mp_obj_t samd_qspiflash_ioctl(mp_obj_t self_in, mp_obj_t cmd_in, mp_obj_t arg_in) {
+    samd_qspiflash_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    mp_int_t cmd = mp_obj_get_int(cmd_in);
+
+    switch (cmd) {
+        case MP_BLOCKDEV_IOCTL_INIT:
+            return MP_OBJ_NEW_SMALL_INT(0);
+        case MP_BLOCKDEV_IOCTL_DEINIT:
+            return MP_OBJ_NEW_SMALL_INT(0);
+        case MP_BLOCKDEV_IOCTL_SYNC:
+            return MP_OBJ_NEW_SMALL_INT(0);
+        case MP_BLOCKDEV_IOCTL_BLOCK_COUNT:
+            return MP_OBJ_NEW_SMALL_INT(self->size / self->sectorsize);
+        case MP_BLOCKDEV_IOCTL_BLOCK_SIZE:
+            return MP_OBJ_NEW_SMALL_INT(self->sectorsize);
+        case MP_BLOCKDEV_IOCTL_BLOCK_ERASE: {
+            samd_qspiflash_erase(mp_obj_get_int(arg_in) * self->sectorsize);
+            // TODO check return value
+            return MP_OBJ_NEW_SMALL_INT(0);
+        }
+        case MP_BLOCKDEV_PROGSIZE:
+            return MP_OBJ_NEW_SMALL_INT(256);
+        default:
+            return mp_const_none;
+    }
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_3(samd_qspiflash_ioctl_obj, samd_qspiflash_ioctl);
+
+STATIC const mp_rom_map_elem_t samd_qspiflash_locals_dict_table[] = {
+    { MP_ROM_QSTR(MP_QSTR_flash_size), MP_ROM_PTR(&samd_qspiflash_size_obj) },
```

Comment:
IMO this method can be removed from all Flash objects.  It's equivalent to ioctl BLOCK_COUNT * BLOCK_SIZE.

Reply:
That method existed already in the initial version of flash_samd.c and the value is read from the JDEC id. So it's just replicated.  
Note that BLOCK_SIZE is also determined from the JDEC id, with a default of 4096 it that's not possible,

Reply:
Removed.